presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/WindowMatcher.java

```diff
@@ -16,28 +16,93 @@
 import com.facebook.presto.sql.planner.plan.WindowNode;
 import com.facebook.presto.sql.tree.FunctionCall;
 import com.google.common.base.MoreObjects;
 import com.google.common.collect.ImmutableList;
 
+import java.util.Comparator;
 import java.util.LinkedList;
 import java.util.List;
+import java.util.Map;
 import java.util.stream.Collectors;
 
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableList;
 import static java.util.Objects.requireNonNull;
 
 final class WindowMatcher
         implements Matcher
 {
+    private final WindowNode.Specification specification;
     private final List<FunctionCall> functionCalls;
 
-    WindowMatcher(List<FunctionCall> functionCalls)
+    WindowMatcher(WindowNode.Specification specification, List<FunctionCall> functionCalls)
     {
+        this.specification = specification;
         this.functionCalls = ImmutableList.copyOf(requireNonNull(functionCalls, "functionCalls is null"));
     }
 
+    private static boolean samePartitionAndOrder(
+            WindowNode.Specification expected,
+            WindowNode.Specification actual)
+    {
+        if (!expected.getPartitionBy().equals(actual.getPartitionBy()) ||
+                !expected.getOrderBy().equals(actual.getOrderBy())) {
+            return false;
+        }
+
+        Map<Symbol, SortOrder> expectedOrderings = expected.getOrderings();
+        Map<Symbol, SortOrder> actualOrderings = actual.getOrderings();
+
+        if (expectedOrderings.size() != actualOrderings.size()) {
+            return false;
+        }
+
+        /*
+         * Ooof. Sad story time.
+         *
+         * Symbols in the plan get decorated with a number if the same symbol is used twice. E.g. if two parts
+         * of a query both refer to a column foo, you'll end up with symbols foo and foo_32 (for example).
+         *
+         * When we verify the orderings in a plan, the orderings take the form of a Map<Symbol, SortOrder>.
+         * Unfortunately, we have no way of knowing a priori what the number affixed to the symbol name, and
+         * even if we did, we'd be relying on implementation details of the SymbolAllocator.
+         *
+         * What's further unfortunate is that we can't just extend Symbol with something that compares equal
+         * to a Symbol with a name that has a unique id affixed. This is because we end up going through
+         * Map.equals, and if you dig deep enough you end up finding that the hashCodes would have to match too.
+         * Since we can't change how Symbol.hashCode is implemented (and wouldn't want to!), we're back to
+         * needing to know the unknowable.
+         *
+         * Another approach to making to WindowNode.Specifications compare with .equals() that doesn't work
+         * is extending Map to do the comparison in a way that doesn't rely on hashCode. That doesn't work
+         * because WindowNode.Specification copies the members of the map you pass to its constructor.
```

Comment:
Another possibility might be to expose an `@VisibleForTesting` constructor for WindowNode.Specification that uses the map that's passed in rather than copying. By doing that, we could probably put a TreeMap in the expected Specification and use a custom comparator.

That would let us get rid of SymbolStem.hashCode, which is inconsistent with SymbolStem.equals.